File: race_mgmt_python/race/race.py
# Class to facillitate races
# Author: Cameron Denton
# Created: 1/14/2025
# Last Modified: 1/14/2025

# Class for running races
class Race:

    # Constructor that takes a list of vehicles and the number of iterations the race is run
    def __init__(self, vehicles, iterations):
        self._vehicles = vehicles
        self._iterations = iterations
    
    # Run the race
    def race(self):
        # Initialize all vehicles
        for vehicle in self._vehicles:
            print(f"{vehicle.Name()}:{vehicle.Type()} is starting")
            start_speed = vehicle.start()
            print(f"Start speed: {start_speed}")
            vehicle.Update_Position(start_speed)
        
        # Run the race for the specified iterations
        for i in range(self._iterations):
            print(f"\nIteration {i+1}")
            for vehicle in self._vehicles:
                print(f"{vehicle.Name()}:{vehicle.Type()} is driving")
                drive_speed = vehicle.drive()
                print(f"Drive speed: {drive_speed}")
                vehicle.Update_Position(drive_speed)
            
            # Determine and print the standings
            standings = sorted(self._vehicles, key=lambda v: v.Position(), reverse=True)
            print("\nStandings:")
            for idx, vehicle in enumerate(standings[:3], start=1):  # Get top 3 vehicles
                print(f"{idx} - {vehicle.Name()} ({vehicle.Type()}): Position {vehicle.Position()}")
            
            print("===========================")

        # Final results
        print("\nFinal Results:")
        final_standings = sorted(self._vehicles, key=lambda v: v.Position(), reverse=True)
        for idx, vehicle in enumerate(final_standings, start=1):
            print(f"{idx} - {vehicle.Name()} ({vehicle.Type()}): Final Position {vehicle.Position()}")

File: race_mgmt_go/internal/race/race.go
// Race management implementation
// Author: Cameron Denton
// Created: 01/14/2025
// Last Modified: 01/14/2025

package race

import (
	"fmt"
	interfaces "race_mgmt_go/internal/race/interfaces"
	"sort"
)

// Race runs a simulation with the provided vehicles for a set number of iterations.
func Race(vehicles []interfaces.Vehicle, iterations int) {
	// Initialize all vehicles
	for _, vehicle := range vehicles {
		startSpeed := vehicle.Start()
		fmt.Printf("Start speed: %d\n", startSpeed)
		vehicle.Update_Position(startSpeed)
	}

	// Run the race for the specified iterations
	for i := 1; i <= iterations; i++ {
		fmt.Printf("\nIteration %d\n", i)

		for _, vehicle := range vehicles {
			fmt.Printf("%s (%s) is driving\n", vehicle.Get_Name(), vehicle.Type())
			driveSpeed := vehicle.Drive()
			fmt.Printf("Drive speed: %d\n", driveSpeed)
			vehicle.Update_Position(vehicle.Get_Position() + driveSpeed)
		}

		// Determine and print the standings
		sort.SliceStable(vehicles, func(i, j int) bool {
			return vehicles[i].Get_Position() > vehicles[j].Get_Position()
		})

		fmt.Println("\nStandings:")
		for idx, vehicle := range vehicles[:3] { // Display top 3 vehicles
			fmt.Printf("%d - %s (%s): Position %d\n", idx+1, vehicle.Get_Name(), vehicle.Type(), vehicle.Get_Position())
		}

		fmt.Println("===========================")
	}

	// Final results
	fmt.Println("\nFinal Results:")
	sort.SliceStable(vehicles, func(i, j int) bool {
		return vehicles[i].Get_Position() > vehicles[j].Get_Position()
	})
	for idx, vehicle := range vehicles {
		fmt.Printf("%d - %s (%s): Final Position %d\n", idx+1, vehicle.Get_Name(), vehicle.Type(), vehicle.Get_Position())
	}
}
